lib/luacrypto_shash.c

```diff
@@ -0,0 +1,289 @@
+	lunatik_free(digest_buf);
+	return 1;
+}
+
+/***
+* Initializes a multi-part hash operation.
+* This must be called before using `update()` or `final()`.
+* @function crypto_shash:init
+* @raise Error on failure.
+*/
+static int luacrypto_shash_init_method(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check(L, 1);
+
+	lunatik_try(L, crypto_shash_init, sdesc);
+	return 0;
+}
+
+/***
+* Updates the hash state with more data.
+* Must be called after `init()`. Can be called multiple times.
+* @function crypto_shash:update
+* @tparam string data The data chunk to add to the hash.
+* @raise Error on failure.
+*/
+static int luacrypto_shash_update(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check(L, 1);
+	size_t datalen;
+	const char *data = luaL_checklstring(L, 2, &datalen);
+
+	lunatik_try(L, crypto_shash_update, sdesc, data, datalen);
+	return 0;
+}
+
+/***
+* Finalizes the multi-part hash operation and returns the digest.
+* Must be called after `init()` and any `update()` calls.
+* @function crypto_shash:final
+* @treturn string The computed digest.
+* @raise Error on failure.
+*/
+static int luacrypto_shash_final(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check(L, 1);
+	unsigned int digestsize = crypto_shash_digestsize(sdesc->tfm);
+	u8 *digest_buf = lunatik_checkalloc(L, digestsize);
+
+	lunatik_try(L, crypto_shash_final, sdesc, digest_buf);
+	lua_pushlstring(L, (const char *)digest_buf, digestsize);
+	lunatik_free(digest_buf);
+	return 1;
+}
+
+/***
+* Combines update and finalization for a multi-part hash operation.
+* Updates the hash state with the given data, then finalizes and returns the digest.
+* `init()` must have been called prior to calling `finup()`.
+* @function crypto_shash:finup
+* @tparam string data The final data chunk.
+* @treturn string The computed digest.
+* @raise Error on failure.
+*/
+static int luacrypto_shash_finup(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check(L, 1);
+	size_t datalen;
+	const char *data = luaL_checklstring(L, 2, &datalen);
+	unsigned int digestsize = crypto_shash_digestsize(sdesc->tfm);
+	u8 *digest_buf = lunatik_checkalloc(L, digestsize);
+
+	lunatik_try(L, crypto_shash_finup, sdesc, data, datalen, digest_buf);
+	lua_pushlstring(L, (const char *)digest_buf, digestsize);
+	lunatik_free(digest_buf);
+	return 1;
+}
+
+/***
+* Exports the internal state of the hash operation.
+* This allows suspending and later resuming a hash calculation via `import()`.
+* Must be called after `init()` and any `update()` calls if part of a multi-step operation.
+* @function crypto_shash:export
+* @treturn string The internal hash state as a binary string.
+* @raise Error on failure (e.g., allocation error).
+*/
+static int luacrypto_shash_export(lua_State *L) {
+	struct shash_desc *sdesc = luacrypto_shash_check(L, 1);
+	unsigned int statesize = crypto_shash_statesize(sdesc->tfm);
+	void *state_udata = lua_newuserdatauv(L, statesize, 0);
```

Comment:
You are returning a userdata not a string.. it's not clear to me how it should be used from Lua